### PR TITLE
nixos/seafile: add persistent user, configurable storage path, and gc service

### PR DESCRIPTION
**Pull request by melvyn2**
## Description of changes
Disable `DynamicUser` for seafile in favor of a persistent configurable service user and group. This allows moving the storage directory out of the systemd-managed `StateDirectory` (previously, external paths could be bind-mounted or symlinked into the state directory, but UID reassignments meant having to manually and slowly `chown` the entire storage). 

This also allows access by other services, such as by a garbage-collection timer also added here.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by melvyn2** [comment]
@greizgh looks you weren't auto-requested by the bot since the module is missing the `meta.maintainer` attr. Would you review this PR?

**Comment by greizgh** [comment]
> @greizgh looks you weren't auto-requested by the bot since the module is missing the `meta.maintainer` attr. Would you please review this PR?

Sorry, completely missed that. This lgtm! Thanks for your contribution, this was deep down on my todo list and I never got the time to look into GC.

**Review by greizgh — APPROVED** [review]

**Comment by melvyn2** [comment]
Added you and @schmittlauch to the maintainers for this module, so we can have the maintainer approval tag hopefully.

File: nixos/modules/services/networking/seafile.nix
Comment (SuperSandro2000) [inline nixos/modules/services/networking/seafile.nix:377]:
```suggestion
            if [[ $installedMajor != $pkgMajor && $installedMinor != $pkgMinor ]]; then
                echo "Server not upgraded yet" >&2
```

File: nixos/modules/services/networking/seafile.nix
Comment (SuperSandro2000) [inline nixos/modules/services/networking/seafile.nix:344]: Can we easily deduplicate more things?

File: nixos/modules/services/networking/seafile.nix
Comment (melvyn2) [reply nixos/modules/services/networking/seafile.nix:377]: Was kept to match the main script, but that isn't too useful I guess.

**Comment by melvyn2** [comment]
Seahub needs to be updated anyways so it can build again, version 10 is using django 3 (marked insecure), and there's no PR yet to update to 11. I'll take a look at this again once that PR's in motion.

**Comment by melvyn2** [comment]
Forgot to include it, update PR is #318727. Switching to draft until that's merged...

**Comment by melvyn2** [comment]
Closing in favor of #318727

**Pull request closed without merging**
